### ISSUE TITLE
Bring DOIs into the core application and refactor to support deposit status

### ISSUE THREAD

**Issue by NateWr** [pkp/pkp-lib#7014]
## Motivation
DOIs are now a settled standard in scholarly publishing. DOIs currently use the `pubId` plugin category, but this was built to make it easier to add pub ids to several object types (submissions, issues, galleys, submission files, chapters), and automatically have the form fields added to the forms.

However, it's become clear that DOI management is best done at the journal level (see https://github.com/pkp/pkp-lib/issues/6062). It has become increasingly difficult to build the desired UIs while maintaining compatiblity with the `pubId` plugin category architecture.

In addition, the `pubId` plugin category architecture does not provide tools for managing the registration and deposit of DOIs and metadata, which have become an important UX concern.

cc @ewhanson, @bozana, @asmecher

## Proposal
We should replace the DOI plugin with DOI functionality that is built into the core application(s), and refactor how DOIs are handled to support registering and depositing them with third-party services.

The following is an overview of the work required:
- [ ] A new database table for DOIs that gives each DOI a unique key and allows status to be tracked at the DOI level.
- [ ] DOIs should be assigned to their objects by their key in the database table, rather than saving DOI information directly into `publication_settings`, `submission_file_settings`, etc. This will make it easier to prevent duplicates and auto-generate DOIs.
- [ ] DOIs should be given a list of status that can be used to manage DOI registration and depositing (see below)
- [ ] By default, DOIs should be generated automatically using the [Cool DOIs](https://blog.datacite.org/cool-dois/). DOI generation patterns should be removed where possible and discouraged where they can't be removed.

### Statuses
DOIs should be assigned one of a few statuses which can be used to track their registration and deposit. These statuses should work regardless of whether the journal is depositing the DOIs manually or using our tools to deposit. They should also work regardless of whether the journal deposits to Crossref, Datacite, or another service.

Suggested statuses:
- **Unregistered: ** No attempt has been made to register this DOI.
- **Submitted:** The DOI has been marked for submission, but has not yet been registered. For example, OJS/OMP/OPS may create a job to register the DOI but the job may still be in the jobs queue.
- **Registered:** The DOI has been successfully registered and deposited in another service. This is to be used if it has been deposited in _any_ third-party service, and whether that deposit has been done through OJS/OMP/OPS or if the DOI has been manually marked as registered.
- **Error:** The DOI could not be registered or deposited in another service because of an error (examples: failed network requests, fatal software error, failed response from Crossref due to invalid data).
- **Stale / Requires Update:** The DOI has already been registered, but the data about this record has changed and any third-party deposit service should be updated about this change.

## Discussion
The following questions are still outstanding:

1. DOIs are stored at the object level (submission, galley, etc), but they are managed and deposited at the submission level. How can we structure the data so that we can make performant searches to identify DOIs that need to be registered, while still displaying the management UI at the submission level? For example, if just one galley of a submission has an `unregistered` DOI but the publication has a `registered` DOI, how can we make it easy and performant to say "get me all submission with an unregistered DOI"? 
2. How can we build the DOI management UI to support additional data (like Crossref event logging) so that any service can record data and it can be exposed in the UI.
3. Would a journal/press/preprint server ever need to deposit to two services at once, eg - Crossref and DataCite?
4. What's the best language to use: "registered", "deposited", "minted"?

**Comment by ajnyga** [comment]
4. `Register` is the right word imo. For example https://www.crossref.org/services/content-registration/ https://support.datacite.org/docs/doi-registration-policy https://www.medra.org/en/guide.htm and https://www.doi.org/registration_agencies.html

**Comment by ewhanson** [comment]
Hi @NateWr,

Here's an initial in-progress look at this work. There are still some core pieces to be finished, but wanted to get your initial thoughts. Thanks!

PRs:

#7350
pkp/ojs#3196
pkp/ui-library#165
pkp/crossref-ojs#4
pkp/orcidProfile#191
pkp/citationStyleLanguage#83

**Comment by NateWr** [comment]
Phew! That was a lot of code! :joy: I've left some comments on the PRs. The following arose from clicking around and testing it out.

- [ ] Move DOI settings to Settings > DIstribution (see below)
- [ ] The big blue deposit button should always deposit any undeposited DOIs, not just the selected ones. You'll need to ping the server to get an accurate count of how many will be deposited.
- [ ] When I marked a bunch registerd, the "registration successful" message didn't appear until after I reloaded the page. This probably uses the old notification system somewhere.
- [ ] The deposit and export buttons shouldn't appear if no registration agency has been set up. And the deposit buttons should only appear if the agency has been configured properly. I'd recommend using a `<notification>` component at the top of the DOI page if agency details are not configured.
- [ ] I was unable to edit the DOI of an already published submission. Got a 403 response when it tried to update the publication.
- [ ] In the issues table of DOIs, there is no "Type" column. That can just say "Issue" for consistency.
- [ ] On a submission with a DOI, if I click "Edit" and then "Save" right away, without making changes, it didn't save anything and lots of output appeared in the console.
- [ ] I have a recommendation for changing how items are shown in the list (see below)
 
## Move DOI settings to Distribution/DOIs
The settings can be moved to a tab under distribution. And the second tab, "regisration", can be made to load the agency's settings form when it is selected. (the form will have to be one of the new forms).
![doi-settings](https://user-images.githubusercontent.com/2306629/135499308-4a96a3d0-7134-41ff-9a6c-64f45c1bc1ca.png)
![doi-agency](https://user-images.githubusercontent.com/2306629/135499323-87a8bf2e-8e60-48cf-ac5d-49aa1b38dcb6.png)

## Simplify display of titles in lists
Make submissions and issues follow the same pattern to simplify the code.
Submissions:
![list-items-submission](https://user-images.githubusercontent.com/2306629/135501793-4b15a03e-ec7f-42b8-b7d3-dd96443b6489.png)
Issues:
![list-items-issues](https://user-images.githubusercontent.com/2306629/135501790-776d247b-32f6-4f3a-8b24-400fb8d60cbe.png)

**Comment by ewhanson** [comment]
Hi @NateWr, I've updated the draft PRs with all the recent changes/revisions. Could you have a look?

In the mean time, I'll prepare the OMP and OPS specific code. Thanks!

**Comment by isgrim** [comment]
Hi @NateWr ,

thanks for working on this. That means one less plugin to bother about.

Does new DOI assignment and registration work with Chapter in OMP as well?

@withanage and  @nongenti worked on  the OMP Datacite plugin to extend the UI for chapters.

**Comment by NateWr** [comment]
@isgrim It's @ewhanson that's done all the work on this so they get the credit. :) If chapter DOI support is in the main branch of OMP right now it will be supported by this work.

**Comment by NateWr** [comment]
@ewhanson two other things that I think are quick and not worth filing separately:

- If an item is deposited, the Edit DOIs button is disabled. But in such cases let's remove it altogether.
- Change the "Issues" filter name to "Assigned to Issue" (a .po key exists for this).

**Comment by ewhanson** [comment]
Hi @NateWr, I've updated OJS and added in support to OPS and OMP. Could you have a look? Thanks!

Here are the new PRs:

pkp/omp#1057
pkp/ops#225
pkp/crossref-ops#8

...along with the existing PRs:
#7350
pkp/ojs#3196
pkp/ui-library#165
pkp/crossref-ojs#4
pkp/orcidProfile#191
pkp/citationStyleLanguage#83

**Comment by NateWr** [comment]
Nice work Erik! I've had a chance to tinker with OJS, OMP and OPS. There were a few teething problems for me, but I tested on brand new installations for each app. It looks like a fair number of bugs have creeped in during the port to other apps. No big deal. I've left a few comments in the pull requests, but most of it is here. Once you're through these I can do a more thorough review.

- I'm not sure if I got something wrong with the setup, but I wasn't able to create an XML export in any app. This was the relevant error:

```
[26-Jan-2022 12:03:23 Europe/London] PHP Warning:  assert(): assert(count($exportFilters) == 1) failed in /home/nate/Projects/pkp/ojs-main/classes/plugins/PubObjectsExportPlugin.inc.php on line 477
[26-Jan-2022 12:03:23 Europe/London] Slim Application Error:
Type: Error
Message: Call to a member function setDeployment() on null
File: /home/nate/Projects/pkp/ojs-main/classes/plugins/PubObjectsExportPlugin.inc.php
Line: 480
```

- I also got an error when I tried to deposit an item that was already submitted. I would expect the UI to maybe tell me if the item I am trying to deposit has already been submitted, or if not then a sensible error message after sending the request.
- Once I got the automated deposit working in the scheduled tasks, the task would create the job. But when I tried to run the jobs with the CLI tool I got this error (maybe related to export problems):

```
	#0 /home/nate/Projects/pkp/ojs-main/plugins/generic/crossref/CrossRefExportPlugin.inc.php(201): APP\plugins\PubObjectsExportPlugin->exportXML()
	#1 /home/nate/Projects/pkp/ojs-main/plugins/generic/crossref/CrossRefPlugin.inc.php(221): CrossRefExportPlugin->exportAndDeposit()
	#2 /home/nate/Projects/pkp/ojs-main/lib/pkp/Jobs/Doi/DepositSubmission.php(63): CrossRefPlugin->depositSubmissions()
```

- When trying to run the scheduled tasks with Acron, there's a bug that prevents the Acron plugin from reloading the scheduled tasks. `PubObjectsExportPlugin::callbackParseCronTab()` assumes there is a scheduledTasks.xml file for each plugin. This then causes an error in `PKPAcronPlugin::_parseCrontab` ("Error parsing scheduled tasks XML file"). I think you'll need to refactor `PubObjectsExportPlugin` to check if the `scheduledTasks.xml` exists and skip the process if it doesn't exist.
- The issue filter on the submission DOI list panel always showed submissions from the same issue no matter which issue I selected in the filter.
- OMP and OPS still have full DOI plugins but I think these are supposed to be removed.

**OMP**
- I could never get a submission to appear in the DOI list panel, whether or not it was published, and whether or not it had assigned DOIs...
- There's a fatal error when trying to create chapter DOIs. Maybe a forgotten foreign key in the install migration (as opposed to the upgrade migration):

```
PHP Fatal error:  Uncaught PDOException: SQLSTATE[23503]: Foreign key violation: 7 ERROR:  insert or update on table "submission_chapters" violates foreign key constraint "submission_chapters_doi_id_foreign"
DETAIL:  Key (doi_id)=(0) is not present in table "dois".
```

- The settings form still has automatic deposit and registration agency options, but these don't exist for OMP right?
- The title of the DOIs list panel is showing as ##submission.list.monographs##.

**OPS**
- I get the following error when exporting. I think it's due to `crossref/filters/filterConfig.xml`:
```
PHP Fatal error:  require_once(): Failed opening required '/home/nate/Projects/pkp/ops-main/plugins/importexport/crossref/filter/PreprintCrossrefXmlFilter.inc.php' (include_path='.:/usr/share/php') in /home/nate/Projects/pkp/ops-main/lib/pkp/includes/functions.inc.php on line 34
```
- I couldn't get a DOI to be assigned when posting a preprint. I think I've configured everything in the settings, but it was never automatically assigned. I always had to manually assign it in the DOI management handler.
- The following appeared in the DOI List Panel: `##plugins.generic.crossref.registrationAgency.name##`

**Enhancements/etc**
- DOIs should be enabled by default after installation. I had to enable them and configure the settings, but we'll probably want them enabled by default on submissions. During our tests, the prefix should be set early so we can test automated DOI creation and stuff. (This can be spun off into a new issue.)
- How do I test DOIs with an error status? Is there a way I can get an error response from Crossref using the sandbox API?

**Comment by ewhanson** [comment]
Hi @natewr, thanks for reviewing this. Most, if not all, of the issues you've mentioned have been taken care of now. 

I've outline the changes and my responses below just to keep track of everything. There will be a few to take a look at together. Thanks!

---

> I'm not sure if I got something wrong with the setup, but I wasn't able to create an XML export in any app. This was the relevant error.

I think we worked through this last week and it was related to the changed Crossref submodule not being moved over.

> I also got an error when I tried to deposit an item that was already submitted. I would expect the UI to maybe tell me if the item I am trying to deposit has already been submitted, or if not then a sensible error message after sending the request.

I haven't directly addressed this as I wasn't able to get an error, but I agree that attempting to redeposit items that have already been scheduled will likely cause problems. I have some ideas on how to approach this, but I'll do a bit more thinking on it and pick your brain on it in our next call.

> Once I got the automated deposit working in the scheduled tasks, the task would create the job. But when I tried to run the jobs with the CLI tool I got this error (maybe related to export problems):

Which CLI tool were you using when you got this error? I imagine if it's the one I modified/worked on, it will need some changes now that all of the Jobs work has been merged. I wasn't able to reproduce the error so hopefully this was related to the plugin submodule issue.

> When trying to run the scheduled tasks with Acron, there's a bug that prevents the Acron plugin from reloading the scheduled tasks. `PubObjectsExportPlugin::callbackParseCronTab()` assumes there is a scheduledTasks.xml file for each plugin. This then causes an error in `PKPAcronPlugin::_parseCrontab` ("Error parsing scheduled tasks XML file"). I think you'll need to refactor `PubObjectsExportPlugin` to check if the `scheduledTasks.xml` exists and skip the process if it doesn't exist.

I've updated the `PubObjectsExportPlugin` to have it check if a `scheduledTasks.xml` file exists rather than assuming it does.

> The issue filter on the submission DOI list panel always showed submissions from the same issue no matter which issue I selected in the filter.

I'm not sure what would be causing this issue and wasn't able to recreate it. If you're still running into it after this round of changes, let's take a look at it on our next call.

> OMP and OPS still have full DOI plugins but I think these are supposed to be removed.

Yep! I've emptied these out but left the locale files for the time being until I do the final locale reconciliation with Weblate before merge.

> I could never get a submission to appear in the DOI list panel, whether or not it was published, and whether or not it had assigned DOIs...

This was for OMP and related to submissions being in the copyediting stage and not production. You suggested we leave this for now and take a look in a UX review.

> There's a fatal error when trying to create chapter DOIs. Maybe a forgotten foreign key in the install migration (as opposed to the upgrade migration):

This is fixed now. It was from coercing `doiId` to int when it should have stayed null instead of 0, i.e. `(int) $chapter->getData('doiId')`

> The settings form still has automatic deposit and registration agency options, but these don't exist for OMP right?

Yes, I've removed it.

> The title of the DOIs list panel is showing as ##submission.list.monographs##.

These have been fixed. 

> I get the following error when exporting. I think it's due to crossref/filters/filterConfig.xml:

I think after looking at this together, it could be attributed to submodule moving/renaming.

> I couldn't get a DOI to be assigned when posting a preprint. I think I've configured everything in the settings, but it was never automatically assigned. I always had to manually assign it in the DOI management handler.

I'm not sure what might be going on here. With a fresh install locally, DOIs are assigned as expected. This is another one to take a look at together.

> The following appeared in the DOI List Panel: ##plugins.generic.crossref.registrationAgency.name##

Fixed.

**Comment by NateWr** [comment]
> The issue filter on the submission DOI list panel always showed submissions from the same issue

> I'm not sure what would be causing this issue and wasn't able to recreate it. 

I just updated again and could still reproduce it. It looks like the `issueIds` query param is getting mangled. When I inspect the request, I see the following  (notice the `[0][]` key in `issueIds`).

```json
{
	"GET": {
		"scheme": "http",
		"host": "localhost:8000",
		"filename": "/publicknowledge/api/v1/submissions",
		"query": {
			"stageIds[]": [
				"0",
				"5"
			],
			"issueIds[0][]": "2",
			"searchPhrase": "",
			"count": "30",
			"offset": "0",
			"_": "1643819300803"
		},
		"remote": {
			"Address": "127.0.0.1:8000"
		}
	}
}
```

When I compare this to the request payload when using the `issueIds` filter in the submissions list, I see a straightforward array:

```json
{
	"GET": {
		"scheme": "http",
		"host": "localhost:8000",
		"filename": "/publicknowledge/api/v1/_submissions",
		"query": {
			"status[]": [
				"4",
				"3",
				"5"
			],
			"issueIds[]": "2",
			"searchPhrase": "",
			"count": "30",
			"offset": "0",
			"_": "1643819238498"
		},
		"remote": {
			"Address": "127.0.0.1:8000"
		}
	}
}
```

This is probably related to how you are compiling this query param in the `DOIListPanel`.

**Comment by ewhanson** [comment]
@NateWr, everything's been updated and is ready for review again. Thanks!

**Comment by NateWr** [comment]
Looks good, @ewhanson. I've done a review and it's all working great. I left a few comments on the pull requests, mostly about getting more code into pkp-lib. I'm happy to merge it as-is, though, so if you want we can file these as separate issues.

A couple more things that you know about but just to flag them up:

- There are still some `// TODO` comments scattered throughout the code. These need to be removed before merge.
- The big locale string swap, which I know you know about.
- The commit messages need to be tidied up. I presume you'll rebase/squash after you've done the locale stuff.
- In OMP, when Automatic DOI Creation is set to "never", it still assigns a DOI when I publish a submission that is in the production stage.

I've also filed a couple new issues to be tackled after merge:

- https://github.com/pkp/pkp-lib/issues/7682
- https://github.com/pkp/pkp-lib/issues/7683

**Comment by ewhanson** [comment]
Hey @NateWr, these should be all ready to go once the test can be re-run (It looks like 7.4 compatibility is broken right now: https://github.com/pkp/pkp-lib/commit/c2c4bbf05d5681ce8ccc597eeeca66f4884209e4#commitcomment-66584171). I had all tests passing and had gone over a final time to double check comments, doc blocks, TODOs, etc. Thanks!

The PRs are:
#7350
pkp/ojs#3196
pkp/ui-library#165
pkp/crossref-ojs#4
pkp/orcidProfile#191
pkp/citationStyleLanguage#83
pkp/omp#1057
pkp/ops#225
pkp/crossref-ops#8

**Comment by NateWr** [comment]
That's great, @ewhanson. I've fixed the PHP 7.4 issue in `main`. Can you rebase again? I know the test failures weren't in your code, but I want to see the 7.4 tests passing on your code before merge.

**Comment by NateWr** [comment]
:tada: All merged! Thanks @ewhanson!
